mm2src/coins/lightning.rs

```diff
@@ -472,6 +477,134 @@ impl LightningCoin {
+
+        let coin = self.clone();
+        let fut = async move {
+            let payment_preimage = PaymentPreimage(preimage);
+            coin.channel_manager.claim_funds(payment_preimage);
+            coin.db
+                .update_payment_preimage_in_db(payment_hash, payment_preimage)
+                .await
+                .error_log_with_msg(&format!(
+                    "Unable to update payment {} information in DB with preimage: {}!",
+                    hex::encode(payment_hash.0),
+                    hex::encode(preimage)
+                ));
+            Ok(TransactionEnum::LightningPayment(payment_hash))
+        };
+        Box::new(fut.boxed().compat())
+    }
+
+    fn validate_swap_payment(&self, input: ValidatePaymentInput) -> ValidatePaymentFut<()> {
+        let payment_hash = try_f!(payment_hash_from_slice(&input.payment_tx)
+            .map_to_mm(|e| ValidatePaymentError::TxDeserializationError(e.to_string())));
+        let payment_hex = hex::encode(payment_hash.0);
+
+        let amt_msat = try_f!(sat_from_big_decimal(&input.amount, self.decimals()));
+
+        let coin = self.clone();
+        let fut = async move {
+            match coin.db.get_payment_from_db(payment_hash).await {
+                Ok(Some(mut payment)) => {
+                    let amount_received = payment.amt_msat;
+                    // Note: locktime doesn't need to be validated since min_final_cltv_expiry should be validated in rust-lightning after fixing the below issue
+                    // https://github.com/lightningdevkit/rust-lightning/issues/1850
+                    // Also, PaymentReceived won't be fired if amount_received < the amount requested in the invoice, this check is probably not needed.
+                    // But keeping it just in case any changes happen in rust-lightning
+                    if amount_received != Some(amt_msat as i64) {
+                        // Free the htlc to allow for this inbound liquidity to be used for other inbound payments
+                        coin.channel_manager.fail_htlc_backwards(&payment_hash);
+                        payment.status = HTLCStatus::Failed;
+                        drop_mutability!(payment);
+                        coin.db
+                            .add_or_update_payment_in_db(payment)
+                            .await
+                            .error_log_with_msg("Unable to update payment information in DB!");
+                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                            "Provided payment {} amount {:?} doesn't match required amount {}",
+                            payment_hex, amount_received, amt_msat
+                        )));
+                    }
+                    Ok(())
+                },
+                Ok(None) => MmError::err(ValidatePaymentError::UnexpectedPaymentState(format!(
+                    "Payment {} should be found on the database",
```

Comment:
Because the error variant is `UnexpectedPaymentState`, I wanted to highlight the expected state in the error message. At this stage (validating a swap payment), the payment info should be in the database. I can modify the message to be
```rust
"Payment {} is not in the database when it should be"
```
What do you think?